### ISSUE TITLE
When expanding phases in Chrome, it seems that somehow too many deletes or keypresses are sent.

### ISSUE THREAD

**Issue by bjohas** [autokey/autokey#587]
## Classification:

Bug

## Reproducibility:

Always

## Version

AutoKey version:  0.95.10

Used GUI (Gtk, Qt, or both):  gtk

Installed via: (PPA, pip3, …). unsure (I think PPA)

Linux Distribution: Ubuntu 20.04

## Summary

When expanding phases in Chrome, it seems that somehow too many keypresses are sent.

The problem was not present until 'recently' (maybe last week or even just last few days). It may be to do with an upgrade of chrome and/or an upgrade of autokey.

## Steps to Reproduce (if applicable)

For example, for **Chrome**, if I have a phrase that expands
```
abc -> alphabet
```
and I type
```
something abc
```
I get
```
sometalphabet
```
I.e., there are too many characters deleted.

I've also found that if I have a script
```
keyboard.send_keys("<ctrl>+<np_page_up>")
```
and bind this to a key, pressing the key seems to send this combination once or twice. I.e., in Chrome, sometimes the above goes to the previous trab, but maybe 1 out of 2-4 presses the combination appears to execute twice, i.e. moving two tabs back.

**Outside chrome** all this works fine.

The two issues may well be separate, but they occurred at the same time, so I'm guessing they are linked.

# Trace

For the first case, the relevant bit seems to be:
```
2021-07-28 11:07:26,056 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2021-07-28 11:07:26,057 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2021-07-28 11:07:26,058 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2021-07-28 11:07:26,059 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2021-07-28 11:07:26,060 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2021-07-28 11:07:26,061 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2021-07-28 11:07:26,062 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
```
The number of backspaces are correct, but in the UI, more than 5 are deleted.

Trace for 2nd case:
```
2021-07-28 11:06:09,556 DEBUG - iomediator - Send via event interface
2021-07-28 11:06:09,556 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
```
However, those errors also occur in expansions in other applications. Maybe the error matters less in other applications, but does matter in Chrome (whcih is always slugging at replacing abbreviations, and where the number of backspaces often dont delete the full appreviation).

Any ideas?

**Comment by ineuw** [comment]
Other users have the same issues, including myself. Until I accepted the reality of Linux X memory management, the interpretive approach of Python, and Autokey. This is in comparison to Windows and Autohotkey. I use both. 

For the time being, 
* The conflicting keyboard shortcuts of the apps and Ubuntu itself should be disabled or avoided. 
* Must add time delay of the minimum of a 1/10th of a second `time.sleep(0.1)` between keyboard to clipboard, and clipboard to keyboard activities, Data manipulation in the clipboard doesn't require delay.

**Comment by bjohas** [comment]
Thanks @ineuw - I've never had the above issue before. I've had issues where autokey doesn't delete enough characters. Likewise on the copy to/from clipboard. However, I've never had the issue of autokey deleting too many characters... Has that happened for you?

Similarly, I've never had an issue where e.g. this
```
keyboard.send_keys("<ctrl>+<np_page_up>")
```
gets sent twice.

So, yes, `time.sleep(0.1)` is a good idea (and I have that in scripts), but this
```
keyboard.send_keys("<ctrl>+<np_page_up>")
```
is a one-line script, without clipboard interaction. Why does it fire twice? And why does this happen all of a sudden?  Also, these errors are really reliable: They always happen. While some of the expansions are a bit hit and miss (for the memory management), the above errors occur reliably: The first issue 100% of the time, the other issue 50% of the time. 

I know that e.g. expansions are more sluggish in Chrome and therefore sometimes characters are missed. E.g. in Terminal expansions are very snappy and don't tend to go wrong. However, the above behaviour has only 'just' started - maybe due to an upgrade in Chrome and/or autokey and/or Ubuntu.

Some more thoughts would be immensely appreciated. The above issues occur so regularly that they are basically making the Chrome/Autokey combination unusable for me. (But then, Chrome on Linux does have a bunch of issues anyway...)

**Comment by ineuw** [comment]
In my simplistic scheme of things, I avoid sending direct commands and use variables instead. Adding delay before send_keys may help. Let me know.

`ctrl_ = "<ctrl>"`
`nppup_ = "<np_page_up>"`
`time.sleep(0.1)`
`keyboard.send_keys(ctrl_ + nppup_)`

**Comment by bjohas** [comment]
Hi @ineuw - thanks. That doesn't make any difference. I suppose I wouldn't expect it to make a difference, as it just inserts a pause before the command is run - however, the command is a single command, so it doesn't depend on sequencing. It's similar with just pasting text: There's no python involved, but I'm just pasting via the clipboard.

Isn't it strange that this only happens in Chrome? Is that maybe a clue?

Btw. the command should be as follows:
```
ctrl_ = "<ctrl>"
nppup_ = "<np_page_up>"
time.sleep(0.1)
keyboard.send_keys(ctrl_ + "+" + nppup_)
```

**Comment by ineuw** [comment]
It may be so, but I am consumed by resolving related issues. Please try
this, it may help you in this instance.

keyboard.send_keys(ctrl_)
time.sleep(test this by varying the value)
keyboard.send_keys(nppup_)

There seem to be an unresolved communication issue between Autokey, Python
and X-Windows. In this case, it's the first two seem to apply. I am
comparing Autohotkey in Windows, to Autokey in Linux and common issues on
the same hardware, a dual boot Windows/Linux Mint C.
Unfortunately, X-windows may play a role in all this.

**Comment by ileixe** [comment]
I am having a similar issue with chrome.

I have alt + v to ctrl + v mapping and when I hit alt +v, it pasted copied contents twice. All the other mappings look the same. `time.sleep()`  did not work just inserting "v".

**Comment by ileixe** [comment]
It looks somehow related to specific keys. For example, `<ctrl>+v` has a problem while `<ctrl>+w` works fine (chrome tab close happens only once).

**Comment by josephj11** [comment]
@ineuw Your last code snippet cannot possibly work. It separates the modifier key keypress from the modified key keypress.

Using variables for substrings might enhance code readability - especially if the variable is used multiple times - but it will have no measurable effect on script execution. Python is sufficiently fast so that variable assignments and references take a negligible amount of time unless they are repeated a great many times in a tight loop.

**Comment by josephj11** [comment]
If this problem really is chrome-specific, that suggests it may not be an AutoKey issue - especially if the traces look good.

Things to try:

See if it's a Chrome-only problem. If it's an underlying library problem, then it should be reproducible in Brave or Vivaldi, both of which use the chromium infrastructure.

Try replacing the AutoKey API calls with calls to the xdo module or by running xdotool externally using subprocess().
Xdotool uses different libraries than we do. If this reproduces the problem, it pretty much exonerates AutoKey. If it works correctly, you have a workaround and the start of the needed information to file an upstream issue with Chrome or whoever provides the libraries we use.

**Comment by ileixe** [comment]
@josephj11 I tried to repro in different apps. Firefox/Vivaldi had the same issue while Emacs/Text editor/libre/gnome terminal office did not.

**Comment by ileixe** [comment]
> Try replacing the AutoKey API calls with calls to the xdo module or by running xdotool externally using subprocess()

Could you elaborate the step a bit? I will follow up

**Comment by ileixe** [comment]
Ok, I found the reason. Recently I updated chrome version 91 to 92. I can find chrome 92 did not accept xdotool command anymore. It may be related to security reasons? 

Anyway I can get autokey working again after I downgraded chrome to 91.

**Comment by josephj11** [comment]
If the problem occurs in Firefox, then it's not specifically an issue with Chrome/chromium because, last I heard, Firefox still uses the Gecko engine which is an unrelated code base (and we should probably change the title of this issue)
We have not updated AutoKey in a long time, but we're hard  at work on 0.96 which is in beta testing now.

Do you still want help with xdotool? Your last comment suggests you were able to try it on your own. If you do, come over to [Gitter](https://gitter.im/autokey/autokey) so we don't clutter up this issue.

It would be interesting to see if this bug can be recreated using autokey-qt. I suspect it will be.

If you want to update Chrome, you would probably have to convert the relevant phrases to scripts where you can finely control how characters get emitted and add delays between actions. That may or may not help in this situation.

**Comment by genehwung** [comment]
I am having the same issue in the upgraded Chrome. Sleep doesn’t work, each “send_keys” with <right> will send right twice, same goes to <control>+<…>.  Unfortunately, this renders AutoKey useless inside Chrome. 

> If the problem occurs in Firefox, then it's not specifically an issue with Chrome/chromium because, last I heard, Firefox still uses the Gecko engine which is an unrelated code base (and we should probably change the title of this issue)

Don’t see this issue under Firefox.
> It would be interesting to see if this bug can be recreated using autokey-qt. I suspect it will be.

No, it does not. 
> If you want to update Chrome, you would probably have to convert the relevant phrases to scripts where you can finely control how characters get emitted and add delays between actions. That may or may not help in this situation.

Could you elaborate?

**Comment by ineuw** [comment]
I apologize. Didn't know what <np_page_up>" was, and didn't investigate it.
I don't assign variables either in such simple scripts, unless I have a
problem and as a step in the process of resolution. Apologies for wasting
everyone's time.

**Comment by ileixe** [comment]
> Do you still want help with xdotool? Your last comment suggests you were able to try it on your own. If you do, come over to Gitter so we don't clutter up this issue.

No, I will use downgraded chrome for a while until they fix the problem. Thanks for the hint to find out it.

**Comment by josephj11** [comment]
So, summarizing to this point:
This appears to affect autokey-gtk and not autokey-qt.

Although not confirmed, it may be a problem with a new release of Chrome. One poster says the problem occurs in Firefox, but another says it does not.

This appears to be a timing/race condition issue because quick applications seem to work fine and the usual suspects (the slow ones) show the issue. This is a bit odd because it probably shouldn't be front end specific.

At the moment, there are one or two workarounds. Using autokey-qt works and writing a script with extra delays may help.

**Comment by josephj11** [comment]
@genehwung When timing is an issue because the receiving application can't keep up with AutoKey emitting actions at machine speed, there is currently no solution when using phrases. I have submitted an enhancement request #566 that may help with this if it is implemented.

When you run into timing issues like this, the only solution is to replace the affected phrases with scripts. Since they are triggered the same way as phrases, this will not affect usage.

(Usually, the first thing to try with phrases is to select the Paste using clipboard (Ctrl+V) option. This fixes most problems, but the OP is also using macros in one example (special tokens enclosed in `<>` brackets) and they are not supported when pasting as opposed to typing.)

When coding a script, you can emit actions of as little as one keypress at a time and add arbitrary delays between actions using calls to `time.sleep()`. These delays are also needed when using the clipboard API for pasting, etc.

To discuss this further, come over to our [support forum](https://groups.google.com/forum/#!forum/autokey-users) where we can get into whatever specifics you need without cluttering up this issue.

**Comment by bjohas** [comment]
Hi @josephj11, 

I'm curious about this "This appears to affect autokey-gtk and not autokey-qt." I assume I cannot use the qt version on Ubuntu, right? As the gnome desktop manager is gtk based? Happy for you to respond via the support forum if that's better!

Many thanks!

**Comment by josephj11** [comment]
You can have both front ends installed at the same time with no issues (unless you're using Arch or a distro like Manjaro which is based on Arch).

If you install the Qt front end, it will pull in a pile of Qt dependencies if you don't already have them from installing something else. However, they should be  marked as "automatic", so if you later uninstall autokey-qt, apt should subsequently tell you that these packages are no longer needed and can be autoremoved.

**Comment by bjohas** [comment]
I'm confused as to why the gtk would not work, but the qt would, but if that's a solution, then I'll just use the qt version. (Are there any downsides?)

I've opened another issue, as the qt version didn't want to run after installation. #590

**Comment by bjohas** [comment]
Incidentally, for those who are interested (slightly off topic): I've temporarily switched from Chrome to Firefox. 
* Firefox is significantly slower than Chrome on Ubuntu 20.04.
* The Chrome builds for Ubuntu 20.04 (or Chrome on Linux) has some long-standing bugs, like around copy/paste in some situation.
Sigh.

**Comment by bjohas** [comment]
Update: The problem seems to have gone away with Chrome Version 92.0.4515.131 (Official Build) (64-bit) [Ubuntu 20.04]. What are others experiencing?

**Comment by josephj11** [comment]
Glad it fixed itself! That's always nice.

TL;DR: There aren't any obvious downsides to using one front end or the other, but ...

As it stands now, the two front ends (-qt and -gtk) are almost completely separate code bases and use two completely different libraries for desktop/GUI interaction.

This has a few consequences.

Different code and different libraries have different bugs and sometimes have different behaviors. We try to eliminate or minimize as many of these as possible.

Because the libraries are  quite different, not all developers know both of them. This means that sometimes something gets fixed or enhanced in one front end but not in the other. Eventually, they converge.

There is a lot of common code between the two and an effort is underway to extract it to one location, but that may take some time because other issues are often more pressing.

I only use autokey-gtk for testing, so I don't generally find any new bugs in it. One reason I don't use it is a small difference between it and the Qt version that annoys me. Left clicking on the Qt panel icon opens the main AutoKey window and I regularly use this feature. Left clicking on the GTK panel icon brings up the context menu. An issue has been opened for this, but, apparently, fixing this in the GTK version is non-trivial. This is an example of why the two versions don't always function identically.
